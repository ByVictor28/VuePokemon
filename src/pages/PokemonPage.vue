<template>
  <h1 v-if="!pokemonSelected">Wait for a second</h1>
  <div v-else>
    <h1>Who is this pokemon?</h1>
    <PokemonImageVue :pokemonId="pokemonSelected.id" :showPokemon="showPokemon"/>
    <PokemonOptionsVue :pokemons="pokemonArr" @selection="checkAnswer($event)" :showResult="showResult"/>    
    <div v-if="showResult">
      <h2>{{result}}</h2>
      <button @click="newGame" class="bg-red-100 px-3 py-1 rounded-lg border-2 border-black hover:bg-red-300">New game</button>
    </div>
  </div>
</template>

<script>

import PokemonImageVue from '../components/PokemonImage.vue'
import PokemonOptionsVue from '../components/PokemonOptions.vue'
import getPokemonOptions from "@/helpers/getPokemonOptions.js"
import {randomInteger} from "../helpers/getPokemonOptions"

export default {
  name:"Pokemon Page",
  
  data(){
    return{
      pokemonArr:[],
      pokemonSelected:null,
      showPokemon:false,
      showResult:false,
      result:""
    }
  },
  methods:{
    async getPokemonArr(){
      this.pokemonArr =await getPokemonOptions()
      const randomPokemon = randomInteger(0,3)
      this.pokemonArr[randomPokemon].correct = true
      this.pokemonSelected = this.pokemonArr[randomPokemon]
    },
    checkAnswer(event){
      this.showPokemon = true
      this.result = event === this.pokemonSelected.id ? `Congratulation pokemon was ${this.pokemonSelected.name}`:`Opss... pokemon was ${this.pokemonSelected.name}`
      this.showResult=true
    },
    newGame(){
      this.pokemonSelected=null
      this.getPokemonArr()  
      this.showPokemon=false
      this.result = ""
      this.showResult = false
    }
  },
  mounted(){
    this.getPokemonArr()
  },

  components:{
    PokemonImageVue,
    PokemonOptionsVue
  }  
}
</script>