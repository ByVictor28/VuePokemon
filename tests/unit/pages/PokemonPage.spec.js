import {shallowMount , mount} from '@vue/test-utils'
import PokemonPage from "@/pages/PokemonPage"
import {pokemons} from "../mocks/pokemons.mock"

describe('PokemonPage component', () => {
  
  let wrapper
  beforeEach(()=>{
    wrapper = shallowMount(PokemonPage)
  })
  //the screenshot is not eveluated becouse there is only the forst div 

  test('must call getPokemonArr when mount', () => {
    
    const  getPokemonArrSpy = jest.spyOn(PokemonPage.methods,"getPokemonArr")
    shallowMount(PokemonPage)
    
    expect(getPokemonArrSpy).toHaveBeenCalledTimes(1)
  })

  test('Snapshot match with pokemons ready', () => {
    const wrapper = mount(PokemonPage,{
      data(){
        return{
          pokemonArr:pokemons,
          pokemonSelected:pokemons[0],
          showPokemon:false,
          showResult:false,
          result:"",
          score:0
        }
      }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('should show PokemonImage and PokemonOptions', () => {
    // WITH MOUNT
    // const wrapper = mount(PokemonPage,{
    //   data(){
    //     return{
    //       pokemonArr:pokemons,
    //       pokemonSelected:pokemons[0],
    //       showPokemon:false,
    //       showResult:false,
    //       result:"",
    //       score:0
    //     }
    //   }
    // })
    

    // const divImage = wrapper.find('[pokemon-test-id="image"]')
    // const divOptions = wrapper.find('[pokemon-test-id="options"]')

    // expect(divImage.exists()).toBeTruthy()
    // expect(divOptions.exists()).toBeTruthy()

    // WITH SHALLOWMAUNT

    const wrapper = shallowMount(PokemonPage,{
      data(){
        return{
          pokemonArr:pokemons,
          pokemonSelected:pokemons[0],
          showPokemon:false,
          showResult:false,
          result:"",
          score:0
        }
      }
    })

    const imageStub = wrapper.find("pokemon-image-vue-stub")
    const optionsStub = wrapper.find("pokemon-options-vue-stub")
    
    expect(imageStub.exists()).toBeTruthy()
    expect(optionsStub.exists()).toBeTruthy()

    expect(imageStub.attributes("pokemonid")).toBe(pokemons[0].id.toString())
    expect(optionsStub.attributes("pokemons")).toBeTruthy()
    
  })
  test('test checkAnswer', async () => {
    const pokemonSelected = pokemons[0]
    const wrapper = shallowMount(PokemonPage,{
      data(){
        return{
          pokemonArr:pokemons,
          pokemonSelected,
          showPokemon:false,
          showResult:false,
          result:"",
          score:0
        }
      }
    })
    await wrapper.vm.checkAnswer(pokemonSelected.id)
    expect(wrapper.find("h2").text()).toBe(`Congratulation pokemon was ${pokemonSelected.name}`)
    console.log(wrapper.vm)
    expect(wrapper.vm.showPokemon).toBe(true)
    expect(wrapper.vm.score).toBe(1)
    
    await wrapper.vm.checkAnswer(10)
    expect(wrapper.vm.result).toBe(`Opss... pokemon was ${pokemonSelected.name}`)
  })
})