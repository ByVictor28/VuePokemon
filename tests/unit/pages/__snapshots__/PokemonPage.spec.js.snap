// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage component Snapshot match with pokemons ready 1`] = `
<div class="bg-red-200 p-4">
  <h1>Who is this pokemon?</h1>
  <h3 class="font-bold text-2xl">Score: 0</h3>
</div>
<div>
  <div class="" pokemon-test-id="image">
    <div class="select-none w-full h-52 flex flex-col justify-center items-center "><img class="filter brightness-0 w-52 h-52 absolute select-none" src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/dream-world/602.svg" alt="Pokemon Shadow">
      <!--v-if-->
    </div>
  </div>
  <div class="flex flex-col items-center" pokemon-test-id="options">
    <ul class="">
      <li class="bg-red-200 my-4 w-36 py-1 hover:bg-red-400 rounded-md cursor-pointer transition-colors"><span class="text-red-400">tynamo</span></li>
      <li class="bg-red-200 my-4 w-36 py-1 hover:bg-red-400 rounded-md cursor-pointer transition-colors"><span class="text-red-400">mr-mime</span></li>
      <li class="bg-red-200 my-4 w-36 py-1 hover:bg-red-400 rounded-md cursor-pointer transition-colors"><span class="text-red-400">mandibuzz</span></li>
      <li class="bg-red-200 my-4 w-36 py-1 hover:bg-red-400 rounded-md cursor-pointer transition-colors"><span class="text-red-400">spheal</span></li>
    </ul>
  </div><!-- <div v-if="showResult"> -->
  <h2></h2>
  <div class="flex gap-4 justify-center"><button class="bg-red-100 px-3 py-1 rounded-lg border-2 border-black hover:bg-red-300">Next pokemon</button><button class="bg-red-100 px-3 py-1 rounded-lg border-2 border-black hover:bg-red-300">New game</button></div>
</div>
`;
